docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,430 @@
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 02 October 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+SecureStore is a [KVStore](../KVStore/KVStore_design.md) based storage solution, providing security features on the stored data, such as encryption, authentication, rollback protection and write once, over an underlying KVStore class.
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instance is instantiated and initialized. 
+
+# System architecture and high-level design
+
+## Design basics
+
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+![SecureStore Layers](./SecureStore_layers.jpg)
+
+As such, it offers all KVStore APIs. Stored data is protected with the following security options (which can be selected using the creation flags at set):  
+
+- Encryption: Data is encrypted using the AES-CTR encryption method. Nonce is randomly generated and each encryption chunk (16 bytes) uses the incrementing counter.   
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data.
+- Rollback protection: (Requires authentication) CMAC is stored in a designated rollback protected storage (also of KVStore type) when set and compared with the stored one at get.
+- Write once: Key can only be stored once and can't be removed. 
+
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce and external RBP key. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
+
+![SecureStore Record](./SecureStore_record.jpg)
+
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP Keys are generated using a 32-bit incremental counter, which is incremented whenever we have a new key. Start value of the counter is achieved by iterating over the rollback protection KVStore (at initialization) and getting the maximal value. 
```

Comment:
start value is a constant when the securestore is empty and later on by iterating ...

Reply:
This is true if use NVStore. If RBP is a KVStore, we will use a string, which can have a prefix (like "RBP12").